<a href="https://colab.research.google.com/github/RoihansLab/Machine-Learning-Projects/blob/main/04_Building_Classification_Model/iris_classification_random_forest.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Building a Classification Model for the iris dataset**

>  Tutorial From : https://youtu.be/XmSlFPDjKdc?si=vCCuHfM7hbbb-5RC





## **1. Import Libraries**

In [89]:
from sklearn import datasets #for iris dataset
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
import joblib

## **2. Load Iris data set**

In [90]:
df = datasets.load_iris()

In [91]:
df

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

## **3. Input Features**

The **iris** data set contains 4 input features and 1 output variable (the class label)

### **3.1. Input features**

In [57]:
print(df.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


### **3.2. Output features**

In [58]:
print(df.target_names)

['setosa' 'versicolor' 'virginica']


In [59]:
print(df.target)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


## **4. Glimpse of Data**

### **4.1. Input Features**

In [60]:
df.data

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

### **4.2. Output variable (the class label)**

In [61]:
df.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

### **4.3. Assignin Input and Output variables**

In [62]:
X = df.data
y = df.target

### **4.4. Let's examine the data dimension**

In [63]:
X.shape

(150, 4)

In [64]:
y.shape

(150,)

## **5. Build Classification Model Using Random Forest**

### Breakoown how can prediction work?

In [65]:
model = RandomForestClassifier(
    bootstrap=True,
    ccp_alpha=0.0,
    class_weight=None,
    criterion='gini',
    max_depth=None,
    max_features='sqrt',
    max_leaf_nodes=None,
    max_samples=None,
    min_impurity_decrease=0.0,
    min_samples_leaf=1,
    min_samples_split=2,
    min_weight_fraction_leaf=0.0,
    n_estimators=100,
    n_jobs=None,
    oob_score=False,
    random_state=None,
    verbose=0,
    warm_start=False
)


warning: di baris model.fit(X, y) inilah tragedi Data Leakage (Kebocoran Kunci Jawaban) terjadi. Modelnya nggak belajar menebak, tapi murni menghafal 150 soal beserta jawabannya langsung (Overfitting parah)

In [66]:
model.fit(X, y)

RandomForestClassifier()

## **6. Feature Importance**

In [67]:
print(model.feature_importances_)

[0.10949149 0.03306203 0.42472884 0.43271764]


## **7. Make Prediction**

In [68]:
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [69]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [70]:
X.shape

(150, 4)

In [71]:
X[0], X[50], X[100]

(array([5.1, 3.5, 1.4, 0.2]),
 array([7. , 3.2, 4.7, 1.4]),
 array([6.3, 3.3, 6. , 2.5]))

In [92]:
print(model.predict([[5.1, 3.5, 1.4, 0.2]]))
print(model.predict([[7. , 3.2, 4.7, 1.4]]))
print(model.predict([[6.3, 3.3, 6. , 2.5]]))

[0]
[1]
[2]


In [93]:
y[0], y[50], y[100]

(np.int64(0), np.int64(1), np.int64(2))

In [94]:
print(model.predict_proba(X[[0, 50, 100]]))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


## **8. Data Split (80/20 Ratio)**

In [75]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [76]:
X_train.shape, y_train.shape

((120, 4), (120,))

In [77]:
X_test.shape, y_test.shape

((30, 4), (30,))

## **9. Rebuilt Random Forest Model**

In [78]:
new_model = RandomForestClassifier()

new_model.fit(X_train, y_train)

RandomForestClassifier()

### **9.1. Performs prediction samples**

In [98]:
print(new_model.predict([[5.1, 3.5, 1.4, 0.2]]))
print(new_model.predict([[7. , 3.2, 4.7, 1.4]]))
print(new_model.predict([[6.3, 3.3, 6. , 2.5]]))

[0]
[1]
[2]


In [99]:
print(new_model.predict_proba([[5.1, 3.5, 1.4, 0.2]]))
print(new_model.predict_proba([[7. , 3.2, 4.7, 1.4]]))
print(new_model.predict_proba([[6.3, 3.3, 6. , 2.5]]))

[[1. 0. 0.]]
[[0. 1. 0.]]
[[0. 0. 1.]]


### **9.2. Performs prediction on the test set**

predicted class label

In [81]:
print(new_model.predict(X_test))

[0 2 1 1 0 1 0 0 2 1 2 2 2 1 0 0 0 1 1 2 0 2 1 2 2 2 1 0 2 0]


In [82]:
print(y_test)

[0 2 1 1 0 1 0 0 2 1 2 2 2 1 0 0 0 1 1 2 0 2 1 2 2 1 1 0 2 0]


In [83]:
print(new_model.predict_proba(X_test))

[[1.   0.   0.  ]
 [0.   0.14 0.86]
 [0.04 0.8  0.16]
 [0.04 0.92 0.04]
 [1.   0.   0.  ]
 [0.   0.85 0.15]
 [1.   0.   0.  ]
 [1.   0.   0.  ]
 [0.   0.   1.  ]
 [0.06 0.88 0.06]
 [0.   0.   1.  ]
 [0.   0.   1.  ]
 [0.   0.   1.  ]
 [0.   0.99 0.01]
 [1.   0.   0.  ]
 [0.99 0.01 0.  ]
 [0.98 0.02 0.  ]
 [0.   1.   0.  ]
 [0.   1.   0.  ]
 [0.   0.38 0.62]
 [1.   0.   0.  ]
 [0.   0.   1.  ]
 [0.   1.   0.  ]
 [0.   0.46 0.54]
 [0.   0.   1.  ]
 [0.   0.03 0.97]
 [0.   1.   0.  ]
 [0.96 0.04 0.  ]
 [0.   0.   1.  ]
 [1.   0.   0.  ]]


## **10. Model Performance**

In [88]:
print(new_model.score(X_test, y_test))

0.9666666666666667


## **11. Save Model**

In [87]:
joblib.dump(new_model, 'iris_classification_RF')

['iris_classification_RF']